### PR TITLE
Add property wrappers

### PR DESCRIPTION
**Pull request by macblazer**

### CODE REVIEW

File: Sources/ManagedAppConfigLib/AppConfigPlain.swift
Comment (tsfischer) [inline Sources/ManagedAppConfigLib/AppConfigPlain.swift:1]: AppConfig.swift and AppConfigPlain.swift have a lot of code (and comments) in common. Is there a way we can share code for less duplication, since they are so simiar?

**Review by tsfischer — CHANGES_REQUESTED** [review]

File: Sources/ManagedAppConfigLib/AppConfigPlain.swift
Comment (macblazer) [reply Sources/ManagedAppConfigLib/AppConfigPlain.swift:1]: I tried to figure that out myself with attempted refactors.  But I don't want to mix the `ObservableObject` conformance and `@StateObject` property wrapper for SwiftUI into the non-SwiftUI version.  Any ideas?

Comment (tsfischer) [reply Sources/ManagedAppConfigLib/AppConfigPlain.swift:1]: Possibly adding the `DynamicProperty` and `ObservableObject` conformance, as well as the `update()` func, in a SwiftUI-specific extension?

There might be wrinkles that prevent that though... and maybe that precludes using the SwiftUI and Plain variants within the same app?

In any case I'm going to go ahead and approve this... but something to think about.

**Review by tsfischer — APPROVED** [review]

**Review by HarryStrandJamf — APPROVED** [review]
I didn't notice any issues, but you might want to wait for mjKosmic's approval since he's used this before.

File: Sources/ManagedAppConfigLib/AppConfigPlain.swift
Comment (macblazer) [reply Sources/ManagedAppConfigLib/AppConfigPlain.swift:1]: The biggest wrinkle to that is getting the `didSet` code onto the `value` property of the Listener for SwiftUI.  There is some possibility to move the `listenTo` function somewhere shared since the `Listener.listenTo()` function is identical for the two `Listener` classes.

The rest of it is just so slightly out of sync that I couldn't find any better way to implement it.  I'm open to any changes around it to make that easier.

It only really matters on iOS 7.0-12.x and tvOS 10.2-12.x because on all newer OS versions (including on all macOS) the SwiftUI support is going to be there anyway whether or not you're actually using SwiftUI views.  So the `@AppConfig` wrapper is the way to go with **any** code (AppKit, UIKit, SwiftUI) if you only need to support modern OSes.

Comment (macblazer) [reply Sources/ManagedAppConfigLib/AppConfigPlain.swift:1]: Maybe I should adjust the README file with some other words around that.  Something like:

> The `@AppConfig` works on any modern OS whether or not you use SwiftUI, and it does trigger SwiftUI redraws if it is used in a SwiftUI view.

And maybe I should rename the `@AppConfigPlain` to `@AppConfigLegacy` or something like that.  And could even move it to its own module within the library.  Anybody newly adopting this Swift Package would presumably do so in order to modernize their code.  Anybody wanting to support older OSes would need to explicitly include the legacy property wrapper.

Comment (tsfischer) [reply Sources/ManagedAppConfigLib/AppConfigPlain.swift:1]: In that case, I wonder if using `available` checks would be preferable... and having the code for AppConfigPlain just be the AppConfig class for pre-SwiftUI OS's. Then when support for these older OS's rolls off, we can just lop off the support here without affecting any user's code (since users would always be using `AppConfig` as the class name).

Comment (macblazer) [reply Sources/ManagedAppConfigLib/AppConfigPlain.swift:1]: I don't think `@available` will let us have two types with the same names.  Because it is type conformances (`DynamicProperty`, `ObservableObject`) and a property wrapper (`@StateObject`) in the implementation differences, I don't see a way to have one named type that does NOT have those things on older OS versions.

I think the best we could do is to have two modules in the package, each declaring a type named `@AppConfig`.  In consuming code you'd import only ONE of the modules and be able to use `@AppConfig` with either the newer implementation or the one with the legacy implementation.

Comment (tsfischer) [reply Sources/ManagedAppConfigLib/AppConfigPlain.swift:1]: Ugh. I feel like there has to be a way to make this happen, but I'm not seeing it now either :(

Comment (macblazer) [reply Sources/ManagedAppConfigLib/AppConfigPlain.swift:1]: Hm.  I tried putting some current `@AppConfig` calls into the unit tests, and forgot about one key thing.  If the SwiftUI-ish `@AppConfig` is used outside of a SwiftUI view context it doesn't update when changes to prefs occur.

So I was wrong earlier.  The `@AppConfig` is only useful in a SwiftUI view.  `@AppConfigPlain` is useful in any code outside of a SwiftUI view, even on modern OSes.

File: Sources/ManagedAppConfigLib/AppConfig.swift
Comment (mjKosmic) [inline Sources/ManagedAppConfigLib/AppConfig.swift:28]: is `self` an optional?

**Review by mjKosmic — APPROVED** [review]

File: Sources/ManagedAppConfigLib/AppConfig.swift
Comment (macblazer) [reply Sources/ManagedAppConfigLib/AppConfig.swift:28]: Yes, because of the `[weak self] _ in` on the previous line.

**Pull request merged**
